doc/case-studies/xycloans/xycloans_types.tla

```diff
@@ -0,0 +1,87 @@
+--------------------------------- MODULE xycloans_types -----------------------
+EXTENDS Variants
+
+(*
+    The relevant parts of the storage that is accessed by the contract.
+    Note that our view of the storage is partial: at any point in time, we only have
+    access to the parts of the storage that are touched by the contract.
+
+    @typeAlias: storage = {
+        self_instance: {
+            FeePerShareUniversal: Int,
+            TokenId: Str
+        },
+        self_persistent: {
+            Balance: Str -> Int,
+            MaturedFeesParticular: Str -> Int,
+            FeePerShareParticular: Str -> Int
+        },
+        token_persistent: {
+            Balance: Str -> Int
+        }
+    };
+
+    The environment of the xycLoans contract that should be
+    produced by Solarkraft from the transaction metadata:
+
+    @typeAlias: env = {
+        current_contract_address: Str,
+        storage: $storage,
+        old_storage: $storage
+    };
+
+    An external contract call:
+
+    @typeAlias: call =
+          Create({ addr: Str })
+        | Initialize({ token: Str })
+        | Deposit({ from: Str, amount: Int })
+        | Borrow({ receiver_id: Str, amount: Int })
+        | UpdateFeeRewards({ addr: Str})
+    ;
+
+    Finally, a transaction is:
+
+    @typeAlias: tx = {
+        env: $env,
+        call: $call,
+        status: Bool
+    };
```

Comment:
Looking at it a second time, it feels a bit weird to have such a deeply-nested record in TLA+ only to emulate the Soroban API.

In the end, you're only using `call` and `status` in the state machine, which could be lifted into TLA+ variables? `env.storage` is already tracked in a state variable, which leaves only `env.current_contract_address`.

Reply:
In general, `env` may contain other things such as the timestamp. So I would prefer to have `env` to be clearly separated from the other things. Do you like to introduce `env`, `call`, and `status` to be three parameters instead of `tx`? They are all essential to a monitor.

Reply:
Let's keep it as is and iterate from there